helix-vcs/src/git.rs

```diff
@@ -0,0 +1,72 @@
+use std::path::Path;
+
+use git_repository::objs::tree::EntryMode;
+use git_repository::{discover, Commit, ObjectId, Repository};
+
+use crate::DiffProvider;
+
+#[cfg(test)]
+mod test;
+
+pub struct Git;
+
+impl DiffProvider for Git {
+    fn get_file_head(&self, file: &Path) -> Option<Vec<u8>> {
+        debug_assert!(file.is_file());
+        debug_assert!(file.is_absolute());
+
+        // discover a repository, requires a directory so we call parent (should not fail but exit gracefully in that case)
+        let repo = discover(file.parent()?).ok()?;
+        let head = repo.head_commit().ok()?;
+        let file_oid = find_file_in_commit(&repo, &head, file)?;
+
+        let file_object = repo.find_object(file_oid).ok()?;
+        Some(file_object.detach().data)
+    }
+}
+
+/// Finds the object that contains the contents of a file at a specific commit.
+fn find_file_in_commit(repo: &Repository, commit: &Commit, file: &Path) -> Option<ObjectId> {
+    let repo_dir = repo.work_dir()?;
+    let rel_path = file.strip_prefix(repo_dir).ok()?;
+    let rel_path_components = byte_path_components(rel_path)?;
+    let tree = commit.tree().ok()?;
+    let tree_entry = tree.lookup_path(rel_path_components).ok()??;
```

Comment:
In the newly released `git-repository` v0.24 there is `lookup_entry_by_path()` which would allow to pass `rel_path` directly, alleviating the need for `byte_path_components()` 🎉.

Reply:
Sadly the new git-repository version raised the MSRV to 1.60 :/ I only raised the MSRV for helix to 1.59 in #3896, because that was still supported by ubuntu and was enough for the previous version of gitoxide. That was perhaps unfortunate timing on my part. However looking at the ubuntu update policy it seems they always package the rust-version required to build firefox. Considering that firefox 105 releases today with dependency on rust 1.60 and ubuntu already has rust 1.60 in testing perhaps the MSRV can be raised again

Reply:
I opened #3913 if that is merged then I can rebase this PR and upgrade gitoxide.
I might open some PRs against gitoxide myself for minor things that I would like for helix and using those requires bumping the msrc anyway

Reply:
As pointed out by @pickfire below, we want to avoid duplicate bstr dependencies so I think I will do the update to git-repository 0.24 in a follow-up PR once the new bstr version is used by all  dependencies.

Reply:
Can we drop a TODO comment here saying what to change in the future once 0.24 is landed?

Reply:
@archseer gave the OK for a duplicate `bstr` dependency in the matrix chat so I was able to update `gitoxide` and use the new functions. Thanks @Byron!